// react-redux-context/lib/context/Provider.js
'use strict';

exports.__esModule = true;
exports.default = exports.StoreContext = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _redux = require('redux');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var StoreContext = exports.StoreContext = _react2.default.createContext();

var Provider = function (_React$Component) {
  _inherits(Provider, _React$Component);

  function Provider(props) {
    _classCallCheck(this, Provider);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this._unsubscribe = null;

    _this._getState = _this._getState.bind(_this);
    _this.state = _this._getState();

    _this._actions = (0, _redux.bindActionCreators)(_this.props.actions, _this.props.store.dispatch);
    _this._dispatch = _this.props.store.dispatch;
    return _this;
  }

  Provider.prototype.componentDidMount = function componentDidMount() {
    var _this2 = this;

    this._unsubscribe = this.props.store.subscribe(function () {
      _this2.setState(function () {
        return _this2._getState();
      });
    });
  };

  Provider.prototype.componentWillUnmount = function componentWillUnmount() {
    this._unsubscribe && this._unsubscribe();
  };

  Provider.prototype.render = function render() {
    var storeData = {
      state: this.state.current,
      actions: this._actions,
      dispatch: this._dispatch
    };

    return _react2.default.createElement(
      StoreContext.Provider,
      { value: storeData },
      this.props.children
    );
  };

  Provider.prototype._getState = function _getState() {
    return {
      current: this.props.store.getState()
    };
  };

  return Provider;
}(_react2.default.Component);

exports.default = Provider;


Provider.propTypes = {
  store: _propTypes2.default.object.isRequired,
  actions: _propTypes2.default.object.isRequired
};
